subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TarTaskOutputPacker.java

```diff
@@ -127,77 +120,82 @@ private void packMetadata(TaskOutputOriginWriter writeMetadata, TarArchiveOutput
         tarOutput.closeArchiveEntry();
     }
 
-    private long pack(Collection<ResolvedTaskOutputFilePropertySpec> propertySpecs, TarArchiveOutputStream tarOutput) {
+    private long pack(Collection<ResolvedTaskOutputFilePropertySpec> propertySpecs, Map<String, Map<String, FileContentSnapshot>> outputSnapshots, TarArchiveOutputStream tarOutput) {
         long entries = 0;
-        for (ResolvedTaskOutputFilePropertySpec spec : propertySpecs) {
+        for (ResolvedTaskOutputFilePropertySpec propertySpec : propertySpecs) {
+            String propertyName = propertySpec.getPropertyName();
+            Map<String, FileContentSnapshot> outputs = outputSnapshots.get(propertyName);
             try {
-                entries += packProperty(spec, tarOutput);
+                entries += packProperty(propertySpec, outputs, tarOutput);
             } catch (Exception ex) {
-                throw new GradleException(String.format("Could not pack property '%s': %s", spec.getPropertyName(), ex.getMessage()), ex);
+                throw new GradleException(String.format("Could not pack property '%s': %s", propertyName, ex.getMessage()), ex);
             }
         }
         return entries;
     }
 
-    private long packProperty(CacheableTaskOutputFilePropertySpec propertySpec, TarArchiveOutputStream tarOutput) throws IOException {
+    private long packProperty(CacheableTaskOutputFilePropertySpec propertySpec, Map<String, FileContentSnapshot> outputSnapshots, TarArchiveOutputStream tarOutput) throws IOException {
         String propertyName = propertySpec.getPropertyName();
-        File outputFile = propertySpec.getOutputFile();
-        if (outputFile == null) {
+        File root = propertySpec.getOutputFile();
+        if (root == null) {
             return 0;
         }
         String propertyPath = "property-" + propertyName;
-        if (!outputFile.exists()) {
+        if (!root.exists()) {
             storeMissingProperty(propertyPath, tarOutput);
             return 1;
         }
         switch (propertySpec.getOutputType()) {
             case DIRECTORY:
-                return 1 + storeDirectoryProperty(propertyPath, outputFile, tarOutput);
+                return storeDirectoryProperty(propertyPath, root, outputSnapshots, tarOutput);
             case FILE:
-                storeFileProperty(propertyPath, outputFile, tarOutput);
+                storeFileProperty(propertyPath, root, tarOutput);
                 return 1;
             default:
                 throw new AssertionError();
         }
     }
 
-    private long storeDirectoryProperty(String propertyPath, File directory, final TarArchiveOutputStream tarOutput) throws IOException {
+    private long storeDirectoryProperty(String propertyPath, File directory, Map<String, FileContentSnapshot> outputSnapshots, final TarArchiveOutputStream tarOutput) throws IOException {
         if (!directory.isDirectory()) {
```

Comment:
This is to check if something went completely bollocks. Maybe we don't need this check at all.